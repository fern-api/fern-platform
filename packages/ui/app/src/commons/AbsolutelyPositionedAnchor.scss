.fern-anchor {
    @apply absolute hidden md:block;

    a {
        @apply flex items-center border-0 relative h-6 rounded-md;
    }

    .fern-anchor-icon {
        @apply absolute left-0 flex h-6 w-6 items-center justify-center rounded-md;
    }

    .fern-anchor-icon:not(.copied) {
        @apply bg-white text-gray-400 ring-1 ring-gray-400/30 backdrop-blur hover:ring-gray-400/60 dark:bg-white/5;
        @apply dark:ring-gray-700/25 dark:hover:ring-white/20;
    }

    .fern-anchor-icon.copied {
        @apply bg-tag-primary ring-border-primary ring-1 backdrop-blur;

        > svg {
            @apply text-accent-primary h-4 w-4;
        }

        @apply dark:bg-tag-primary-dark dark:ring-border-primary-dark;
    }
}
