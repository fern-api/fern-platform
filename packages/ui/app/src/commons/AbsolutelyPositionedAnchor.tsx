import classNames from "classnames";
import Link from "next/link";
import { LinkIcon } from "./icons/LinkIcon";

export declare namespace AbsolutelyPositionedAnchor {
    export interface Props {
        route: string;
        verticalPosition: "center" | "default";
    }
}

/**
 * Can only be used with a parent div that has `position` set to `"relative"`.
 */
export const AbsolutelyPositionedAnchor: React.FC<AbsolutelyPositionedAnchor.Props> = ({ verticalPosition, route }) => {
    return (
        <div
            className={classNames(
                "absolute -left-[calc(0.875rem+0.375rem*2)] flex items-center justify-center px-1.5 py-1 opacity-0 hover:opacity-100 group-hover/anchor-container:opacity-100 z-10",
                {
                    "top-2.5": verticalPosition === "default",
                    "top-0 bottom-0": verticalPosition === "center",
                }
            )}
        >
            {/* <button onClick={onClick}>
                <LinkIcon className="t-muted hover:text-text-primary-light hover:dark:text-text-primary-dark h-3.5 w-3.5 transition" />
            </button> */}
            <Link href={route} shallow={true} scroll={false} replace={true}>
                <LinkIcon className="t-muted hover:text-text-primary-light hover:dark:text-text-primary-dark h-3.5 w-3.5 transition" />
            </Link>
        </div>
    );
};
