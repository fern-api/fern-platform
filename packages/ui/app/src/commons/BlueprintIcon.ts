import dynamic from "next/dynamic";

// TODO: delete blueprint icons to reduce bundle size
export const BlueprintIcon = dynamic(() => import("@blueprintjs/core").then(({ Icon }) => Icon), {
    ssr: false,
});
