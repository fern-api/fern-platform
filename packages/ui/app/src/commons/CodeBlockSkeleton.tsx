import { type Theme } from "@fern-ui/theme";
import classNames from "classnames";
import React from "react";
import { Prism as SyntaxHighlighter } from "react-syntax-highlighter";
import * as prism from "react-syntax-highlighter/dist/cjs/styles/prism";

type CodeBlockSkeletonProps = {
    className?: string;
    theme?: Theme;
    language: string;
    content: string;
    usePlainStyles?: boolean;
    style?: React.CSSProperties;
};

export const CodeBlockSkeleton: React.FC<CodeBlockSkeletonProps> = ({
    className,
    theme,
    language,
    content,
    usePlainStyles = false,
    style,
}) => {
    return (
        <div
            className={classNames(
                "bg-gray-100/90 dark:bg-gray-950/90",
                "typography-font-code-block",
                {
                    "w-full border-l border-r border-b rounded-bl-lg rounded-br-lg border-border-default-light dark:border-border-default-dark":
                        !usePlainStyles,
                },
                className
            )}
            style={style}
        >
            <SyntaxHighlighter
                style={theme === "dark" ? prism.vscDarkPlus : prism.oneLight}
                customStyle={{
                    width: "100%",
                    overflowX: "auto",
                    margin: 0,
                    paddingRight: 16,
                    paddingLeft: 16,
                    paddingBottom: 20,
                    fontSize: 12,
                    background: "unset",
                    backgroundColor: "unset",
                    fontFamily: "inherit",
                }}
                codeTagProps={{
                    style: {
                        background: "unset",
                        fontFamily: "unset",
                        fontSize: "unset",
                    },
                }}
                language={language}
                PreTag="pre"
            >
                {content}
            </SyntaxHighlighter>
        </div>
    );
};
