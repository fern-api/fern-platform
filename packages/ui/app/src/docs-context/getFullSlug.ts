import { type DocsInfo, type GetFullSlugOpts } from "./DocsContext";

export function getFullSlug(
    slug: string,
    versionSlug: string,
    activeTabSlug: string | undefined,
    docsType: DocsInfo["type"],
    isOnDefaultVersion: boolean,
    opts?: GetFullSlugOpts
): string {
    const { omitVersionSlug = false, omitTabSlug = false } = opts ?? {};
    const parts: string[] = [];
    if (!omitVersionSlug && docsType === "versioned" && !isOnDefaultVersion && versionSlug) {
        parts.push(`${versionSlug}/`);
    }
    if (!omitTabSlug && activeTabSlug != null) {
        parts.push(`${opts?.tabSlug ?? activeTabSlug}/`);
    }
    parts.push(slug);
    return parts.join("");
}
