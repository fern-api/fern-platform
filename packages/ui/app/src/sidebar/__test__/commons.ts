export interface TabbedVersionedFixture {
    type: "tabbed-versioned";
    name: string;
    revision: number | string;
    activeVersionId: string;
    activeTabSlug: string;
}

export interface TabbedUnversionedFixture {
    type: "tabbed-unversioned";
    name: string;
    revision: number | string;
    activeTabSlug: string;
}

export interface UntabbedVersionedFixture {
    type: "untabbed-versioned";
    name: string;
    revision: number | string;
    activeVersionId: string;
}

export interface UntabbedUnversionedFixture {
    type: "untabbed-unversioned";
    name: string;
    revision: number | string;
}

export type Fixture =
    | TabbedVersionedFixture
    | TabbedUnversionedFixture
    | UntabbedVersionedFixture
    | UntabbedUnversionedFixture;
