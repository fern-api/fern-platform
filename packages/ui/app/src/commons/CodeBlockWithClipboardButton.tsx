import classNames from "classnames";
import React from "react";
import { useDocsContext } from "../docs-context/useDocsContext";
import { CodeBlockSkeleton } from "./CodeBlockSkeleton";
import { CopyToClipboardButton } from "./CopyToClipboardButton";

type CodeBlockWithClipboardButtonProps = {
    className?: string;
    language: string;
    content: string;
    clipboardButtonDistanceFromEdges: "sm" | "lg";
};

export const CodeBlockWithClipboardButton: React.FC<CodeBlockWithClipboardButtonProps> = ({
    className,
    language,
    content,
    clipboardButtonDistanceFromEdges,
}) => {
    const { theme } = useDocsContext();
    return (
        <div className={classNames("relative mb-5 w-full", className)}>
            <div className="border-border-default-light dark:border-border-default-dark flex h-4 rounded-t-lg border-x border-t bg-gray-100/90 px-3 dark:bg-gray-950/90" />
            <CopyToClipboardButton
                className={classNames("absolute right-3 top-3 ml-auto", {
                    "right-3 top-3": clipboardButtonDistanceFromEdges === "sm",
                    "right-4 top-4": clipboardButtonDistanceFromEdges === "lg",
                })}
                content={content}
            />
            <CodeBlockSkeleton theme={theme} language={language} content={content} />
        </div>
    );
};
