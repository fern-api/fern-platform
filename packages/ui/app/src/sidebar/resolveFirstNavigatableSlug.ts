import { getSlugForFirstNavigatableEndpointOrWebhook, UrlPathResolver } from "@fern-ui/app-utils";
import { DocsContextValue } from "../docs-context/DocsContext";

export type ResolveResult =
    | {
          slugToNavigate: string;
          pathToPush: string;
      }
    | undefined;

export async function resolveFirstNavigatableSlug(
    slug: string,
    resolver: UrlPathResolver,
    context: Pick<DocsContextValue, "docsDefinition" | "getFullSlug">
): Promise<ResolveResult> {
    const { docsDefinition, getFullSlug } = context;
    const resolvedUrlPath = await resolver.resolveSlug(slug);
    if (resolvedUrlPath?.type === "apiSubpackage") {
        const apiId = resolvedUrlPath.apiSection.api;
        const apiDefinition = docsDefinition.apis[apiId];
        if (apiDefinition == null) {
            return undefined;
        }
        const slugToNavigate = getSlugForFirstNavigatableEndpointOrWebhook(
            apiDefinition,
            [resolvedUrlPath.slug],
            resolvedUrlPath.subpackage
        );
        if (slugToNavigate != null) {
            return {
                slugToNavigate,
                pathToPush: `/${getFullSlug(slugToNavigate)}`,
            };
        }
    }
    return undefined;
}
