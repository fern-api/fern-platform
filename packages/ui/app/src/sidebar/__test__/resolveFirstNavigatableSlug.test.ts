import {
    assertIsUnversionedUntabbedNavigationConfig,
    assertIsVersionedNavigationConfig,
    UrlPathResolver,
} from "@fern-ui/app-utils";
import { loadDocsDefinition } from "@fern-ui/test-utils";
import { getFullSlug } from "../../docs-context/getFullSlug";
import { resolveFirstNavigatableSlug, type ResolveResult } from "../resolveFirstNavigatableSlug";
import { type Fixture } from "./commons";

const FIXTURES: Fixture[] = [
    {
        type: "untabbed-versioned",
        name: "primer",
        revision: 1,
        activeVersionId: "v2.2",
    },
];

describe("resolveFirstNavigatableSlug()", () => {
    for (const fixture of FIXTURES) {
        it(`${fixture.name}-${fixture.revision}`, async () => {
            const docsDefinition = loadDocsDefinition(fixture.name, fixture.revision);
            if (docsDefinition == null) {
                throw new Error("Fixture docs definition does not exist.");
            }

            const navigationConfig = docsDefinition.config.navigation;
            const navigatablesByApiId: Record<string, ResolveResult[]> = {};
            await Promise.all(
                Object.values(docsDefinition.apis).map(async (apiDef) => {
                    let activeVersionSlug: string | undefined;
                    let items;
                    if (fixture.type === "untabbed-versioned") {
                        assertIsVersionedNavigationConfig(navigationConfig);
                        const versionConfigData = navigationConfig.versions.find(
                            (v) => v.version === fixture.activeVersionId
                        );
                        if (versionConfigData == null) {
                            throw new Error(
                                `Expected version config data for version "${fixture.activeVersionId}" to be defined.`
                            );
                        }
                        assertIsUnversionedUntabbedNavigationConfig(versionConfigData.config);
                        items = versionConfigData.config.items;
                        activeVersionSlug = versionConfigData.urlSlug;
                    } else {
                        // TODO: Implement after new resolver
                        throw new Error("Not yet implemented");
                    }

                    const activeVersionSlugConst = activeVersionSlug;

                    // TODO: Replace with the new resolver to complete the tests
                    const resolver = new UrlPathResolver({
                        items,
                        loadApiDefinition: (id) => docsDefinition.apis[id],
                        loadApiPage: (id) => docsDefinition.pages[id],
                        loadMdxCompiler: async () => ({
                            serialize: async () => ({
                                scope: {},
                                compiledSource: "",
                                frontmatter: {},
                            }),
                        }),
                    });

                    await Promise.all(
                        Object.values(apiDef.subpackages).map(async (s) => {
                            const result = await resolveFirstNavigatableSlug(s.urlSlug, resolver, {
                                docsDefinition,
                                getFullSlug: (slug) =>
                                    getFullSlug(slug, activeVersionSlugConst, undefined, "versioned", true),
                            });
                            if (navigatablesByApiId[apiDef.id] == null) {
                                navigatablesByApiId[apiDef.id] = [];
                            }
                            navigatablesByApiId[apiDef.id]?.push(result);
                        })
                    );
                })
            );
            expect(navigatablesByApiId).toMatchSnapshot();
        }, 90_000);
    }
});
