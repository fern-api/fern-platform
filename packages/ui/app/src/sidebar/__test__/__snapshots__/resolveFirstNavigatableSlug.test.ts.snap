// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveFirstNavigatableSlug() primer-1 1`] = `
{
  "67c83aa4-e686-4f16-b02d-4c27a6e92f91": [
    undefined,
    undefined,
  ],
  "73803ef9-1d14-46c9-9c0b-ece23481f286": [
    undefined,
    undefined,
    undefined,
  ],
  "95ddda06-15e0-44ae-8e19-3131ee929616": [
    undefined,
    undefined,
    undefined,
  ],
  "d0eefb20-05d3-41e0-af7a-29afd2b1694e": [
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
  ],
}
`;
